_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+fi
+
+IS_SIMULATOR=0
+if [ "${LLVM_TARGET_TRIPLE_SUFFIX-}" = "-simulator" ]; then
+  IS_SIMULATOR=1
+fi
+
+for arch in $ARCHS; do
+  case "$arch" in
+    x86_64)
+      if [ $IS_SIMULATOR -eq 0 ]; then
+        echo "Building for x86_64, but not a simulator build. What's going on?" >&2
+        exit 2
+      fi
+
+      # Intel iOS simulator
+      export CFLAGS_x86_64_apple_ios="-target x86_64-apple-ios"
+      $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target x86_64-apple-ios
+      ;;
+
+    arm64)
+      if [ $IS_SIMULATOR -eq 0 ]; then
+        # Hardware iOS targets
+        $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target aarch64-apple-ios
+      else
+        $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target aarch64-apple-ios-sim
+      fi
+  esac
+done
+```
+
+It's a bit long, but not too complex:
+It selects the cargo profile to use (that is whether to pass `--release` or not),
+sets up the `LIBRARY_PATH` if necessary and finally compiles the Rust library for the selected target.
+Xcode passes the architectures to build in `ARCHS`.
+It's either `x86_64` for simulator builds on Intel Mac hardware or `arm64`.
+If it's `arm64` it can be either the simulator or an actual hardware target.
+Those differ, but we can know which is which from what's in `LLVM_TARGET_TRIPLE_SUFFIX` and select the right Rust target.
+
+## The code parts
+
+We now have an Xcode project that builds our Rust library and links against it.
+We now need to use this library!
+
+Swift speaks Objective-C, which is an extension to C,
+but we need to tell it about the things available.
+In C land that's done with a header.
+Let's create a new file, select the "Header File" template and name it `FfiBridge.h`.
+This will create a new file with this content:
+
+```c
+#ifndef FfiBridge_h
+#define FfiBridge_h
+
+
+#endif /* FfiBridge_h */
+```
+
+Here we need to add the definition of our function.
+As a reminder this is its definition in Rust:
+
+```rust
+extern "C" fn shipping_rust_addition(a: c_int, b: c_int) -> c_int;
+```
+
+This translates to the following in C:
+
+```c
+int shipping_rust_addition(int a, int b);
+```
+
+Add that line between the `#define` and `#endif` lines.
+Xcode doesn't know about that file yet, so once more into the `Build Settings` of the target.
+Search for `Objective-C Bridging Header` and set it to `$(PROJECT_DIR)/ShippingRust/FfiBridge.h`.
+In `Build Phases` add a new `Header Phase`.
+There you add the `FfiBridge.h` as well.
+
+If it now all compiles we're finally ready to use our Rust library.
+
+Open up `ContentView.swift` and change the code to call your Rust library:
+
+```
+struct ContentView: View {
+    var body: some View {
+        Text("Hello, world! \(shipping_rust_addition(30, 1))")
+            .padding()
+    }
+}
+```
+
+We simply interpolate the result of `shipping_rust_addition(30, 1)` into the string displayed.
```

Comment:
TIL that "string interpolation" is a real phrase even though it has nothing to do with its mathematical meaning. Thanks, I hate it.